src/dashboard/src/components/accounts/signals.py

```diff
@@ -1,11 +1,83 @@
 from __future__ import absolute_import
 
+import logging
+
 from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.dispatch import receiver
 from django_auth_ldap.backend import populate_user
+from django_cas_ng.signals import cas_user_authenticated
 
 from components.helpers import generate_api_key
 
+logger = logging.getLogger("archivematica.dashboard")
+
+
+def _cas_user_is_administrator(cas_attributes):
+    """Determine if new user is an administrator from CAS attributes.
+
+    :param cas_attributes: Attributes dict returned by CAS server.
+
+    :returns: True if expected value is found, otherwise False.
+    """
+    ADMIN_ATTRIBUTE = settings.CAS_ADMIN_ATTRIBUTE
+    ADMIN_ATTRIBUTE_VALUE = settings.CAS_ADMIN_ATTRIBUTE_VALUE
+    if (ADMIN_ATTRIBUTE is None) or (ADMIN_ATTRIBUTE_VALUE is None):
+        logger.error(
+            "Error determining if new user is an administrator. Please "
+            "be sure that CAS settings AUTH_CAS_ADMIN_ATTRIBUTE and "
+            "AUTH_CAS_ADMIN_ATTRIBUTE_VALUE are properly set."
+        )
+        return False
+
+    # CAS attributes are a dictionary. The value for a given key can be
+    # a string or a list, so our approach for checking for the expected
+    # value takes that into account.
+    ATTRIBUTE_TO_CHECK = cas_attributes.get(ADMIN_ATTRIBUTE)
+    if isinstance(ATTRIBUTE_TO_CHECK, list):
+        if ADMIN_ATTRIBUTE_VALUE in ATTRIBUTE_TO_CHECK:
+            return True
+    elif isinstance(ATTRIBUTE_TO_CHECK, str):
+        if ATTRIBUTE_TO_CHECK == ADMIN_ATTRIBUTE_VALUE:
+            return True
+    return False
+
+
+@receiver(cas_user_authenticated)
+def cas_user_authenticated_callback(sender, **kwargs):
+    """Set user.is_superuser based on CAS attributes.
+
+    When a user is authenticated, django_cas_ng sends the
+    cas_user_authenticated signal, which includes any attributes
+    returned by the CAS server during p3/serviceValidate.
+
+    When the CAS_CHECK_ADMIN_ATTRIBUTES setting is enabled, we use this
+    receiver to set user.is_superuser to True if we find the expected
+    key-value combination configured with CAS_ADMIN_ATTRIBUTE and
+    CAS_ADMIN_ATTRIBUTE_VALUE in the CAS attributes, and False if not.
+
+    This check happens for both new and existing users, so that changes
+    in group membership on the CAS server (e.g. a user being added or
+    removed from the administrator group) are applied in Archivematica
+    on the next login.
+    """
+    if not settings.CAS_CHECK_ADMIN_ATTRIBUTES:
+        return
+
+    username = kwargs.get("user")
+    attributes = kwargs.get("attributes")
+
+    if not attributes:
+        return
+
+    User = get_user_model()
+    user = User.objects.select_for_update().get(username=username)
+    is_administrator = _cas_user_is_administrator(attributes)
+    if user.is_superuser != is_administrator:
+        user.is_superuser = is_administrator
+        user.save()
```

Comment:
Thanks for doing this! Looking at it again I'm realizing that we're introducing a couple of new issues,
* The lock is retained longer that we need before _cas_user_is_administrator is executed within the transaction window, but you could easily avoid that by moving up that call, and
* I think that we're not freeing the lock - we'd need to wrap the database-related code using `with transaction.atomic: ...`.

And we don't have to do this at all but I was wondering if this would also be a good case for [conditional expressions](https://docs.djangoproject.com/en/1.8/ref/models/conditional-expressions/#conditional-update) and if that would be a better way to solve it.

Reply:
Yes, good point! In the past I've relied on Django managing transactions for me with autocommit. A reading of the Django [database transactions](https://docs.djangoproject.com/en/1.11/topics/db/transactions/) docs was very instructive.

I've moved `_cas_user_is_administrator` and added the `transaction.atomic` context manager, which should resolve the new issues.

Conditional updates are new to me - thanks for pointing it out! From my reading, they seem most appropriate for cases where an update is required and the trick is to determine which value to set. I think this case is somewhat different, where most logins after the first will not require updates to the user instance, since updates will only be necessary if the administrator attribute is added or removed for a user on the CAS server (which I assume happens rarely). So we want to check to see if our application state is out of date and update only if necessary.